File: 78-subsets/78-subsets.py
class Solution(object):
    def subsets(self, nums):
        res = []
        
        curr = []
        def dfs(i):
            if(i>=len(nums)):
                res.append(copy.copy(curr))
                return
            
            curr.append(nums[i])
            dfs(i+1)
            
            curr.pop()
            dfs(i+1)
        
        dfs(0)
        return(res)
            
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        

File: 78-subsets/78-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> answer = new ArrayList<>();
        List<Integer> ans = new ArrayList<>();
        backtrack(0,nums,ans,answer);
        return answer;
    }
    
    public void backtrack(int index,int[] nums,List<Integer> ans,List<List<Integer>> answer) {
        if(index>nums.length) {
            return;
        }
        answer.add(new ArrayList<>(ans));
        for(int i=index;i<nums.length;i++) {
            ans.add(nums[i]);
            backtrack(i+1,nums,ans,answer);
            ans.remove(ans.size()-1);
        }
        return;
    }
}